ext/semian/sliding_window.c

```diff
@@ -0,0 +1,485 @@
+    resize_window(res->sem_id, res->shmem, error_threshold);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return self;
+}
+
+VALUE
+semian_simple_sliding_window_size(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = RB_INT2NUM(res->shmem->length);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_resize_to(VALUE self, VALUE new_size)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval = Qnil;
+
+  int new_max_size = RB_NUM2INT(new_size);
+  if (new_max_size < 1) {
+    rb_raise(rb_eArgError, "cannot resize to %d", new_max_size);
+  }
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = resize_window(res->sem_id, res->shmem, new_max_size);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_max_size_get(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = RB_INT2NUM(res->shmem->max_size);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_max_size_set(VALUE self, VALUE new_size)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  int new_max_size = RB_NUM2INT(new_size);
+  if (new_max_size < 1) {
+    rb_raise(rb_eArgError, "max_size must be positive");
+  }
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = resize_window(res->sem_id, res->shmem, new_max_size);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_values(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = rb_ary_new_capa(res->shmem->length);
+    for (int i = 0; i < res->shmem->length; ++i) {
+      int index = (res->shmem->start + i) % res->shmem->max_size;
+      int value = res->shmem->data[index];
+      rb_ary_store(retval, i, RB_INT2NUM(value));
+    }
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_last(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    int index = (res->shmem->start + res->shmem->length - 1) % res->shmem->max_size;
+    retval = RB_INT2NUM(res->shmem->data[index]);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_clear(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+
+  sem_meta_lock(res->sem_id);
+  {
+    dprintf("Clearing sliding window");
+    res->shmem->length = 0;
+    res->shmem->start = 0;
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return self;
+}
+
+// Handy for debugging the sliding window, but too noisy for regular debugging.
+/*
+static void dprint_window(semian_simple_sliding_window_shared_t *window)
```

Comment:
Maybe add an `#ifdef` around this or something so you don't have to leave the code commented out?

Reply:
I don't see the difference, because the compiler won't see it anyways, but sure.